### PR TITLE
Advanced DH support

### LINKED ISSUE #498: Implement DH functionality
This is one of the types of keys that can be loaded through PKey. Follow-up of #234.

Needed: key generation, getting key parameters, relevant crypto operations

### PR DESCRIPTION
**Pull request by wiomoc**
This implements the functionality to generate the keys needed and use them for key derivation
closes #498

### CODE REVIEW

**Review by stbuehler — CHANGES_REQUESTED** [review]
Taking or returning 3 values of the same type in a function seems always bad to me, because you are gonna screw up the order eventually. Perhaps this should be wrapped into structs (one for owning them, one for borrowed values)?

You also should add some doc comments for the functions in the `openssl` crate (but not `openssl-sys`); usually a short text what it does and a reference to the openssl man page.

File: openssl-sys/src/dh.rs [hunk 1]
```diff
@@ -16,4 +27,15 @@ extern "C" {
...
+    #[cfg(any(ossl110, libressl273))]
+    pub fn DH_get0_pqg(
+        dh: *const DH,
+        p: *mut *const BIGNUM,
+        q: *mut *const BIGNUM,
+        g: *mut *const BIGNUM,
+    );
+
+    #[cfg(ossl102)]
+    pub fn DH_get0_key(dh: *const DH, pub_key: *mut *const BIGNUM, priv_key: *mut *const BIGNUM);
```
[the remark this answers is not in the document]
Comment (stbuehler) [reply openssl-sys/src/dh.rs:39]: Also `DH_get0_key` got only added in `ossl110` (added in https://github.com/openssl/openssl/commit/0aeddcfa61250a6c474c4f8b3533772a63192f1b#diff-d3cc027864e7dec95eea6555d30d8b63 and constified in https://github.com/openssl/openssl/commit/fd809cfdbd6e32b6b67b68c59f6d55fbed7a9327#diff-d3cc027864e7dec95eea6555d30d8b63).

**Review by stbuehler — COMMENTED** [review]
Mostly looking good to me now.

I think removing `Result<...>` from accessors was the right choice; those shouldn't fail with a consistent state, which the rust types should ensure.

We'll see what sfackler thinks about `DhParamsRef`; perhaps another alternative to returning a tuple would be standalone accessors (`prime_p`, `prime_q`, `generator`?).

The local `DH_get0_pqg` implementation doesn't handle returning only partial data yet like `DH_get0_key` (the openssl does I think); as long as this isn't used it probably doesn't matter.

Tests still fail as `transpose` wasn't stable in rust 1.31.0 (the current min version; I think sfackler considers raising it breaking the API).

File: openssl/src/dh.rs [hunk 2]
```diff
@@ -106,9 +197,50 @@ impl Dh<Params> {
...
+impl<T> Dh<T>
+where
+    T: HasPublic,
+{
+    /// Gets the public key from the DH instance.
+    ///
+    /// This corresponds to [`DH_get0_key`].
+    ///
+    /// [`DH_get0_key`]: https://www.openssl.org/docs/man1.1.0/crypto/DH_get0_key.html
+    pub fn get_public_key(&self) -> &BigNumRef {
```
Comment (sfackler) [inline openssl/src/dh.rs:209]: Rust conventions avoid the `get_` prefix, so this would just be called `public_key`.

**Comment by sfackler** [comment]
I think it's probably best to follow the convention used by e.g. the `Rsa` type and have separate accessors for the different components.

While I do in crease the MSRV over time, I like to keep it where it is where possible. In this case it's easy to work around the lack of transpose so I'd just not use the method.

**Pull request merged**